### PR TITLE
New changes in persistent event store

### PR DESCRIPTION
**Pull request by Praveenrajmani**
## Description
This PR does two things
- Removes memory store. We will not persist events in memory anymore, if `queueDir` is not set.
- Orders the events before replaying to the broker.

## Motivation and Context
We will not persist events in memory store any more. And we preserve the order of events while replaying.

## Regression
No

## How Has This Been Tested?

- Enable queueStore in the mqtt config
- Start minio without an active mqtt broker.
- Bring down the mqtt broker ( can use activemq for testing )
- Start sending the events. The events will persist in the queueDir.
- Start the broker, the persisted events will be replayed.

Note - The server will fail to set the config if the queueDir is not configured and the broker connection is inactive.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Praveenrajmani** [comment]
Will incorporate this change in other event persistence PRs as well. after this gets merged.

File: pkg/event/target/queuestore.go [hunk 1]
```diff
@@ -154,17 +155,28 @@ func (store *QueueStore) del(key string) error {
...
 // lockless call.
-func (store *QueueStore) listN(n int) []string {
+func (store *QueueStore) list() []string {
+	var names []string
 	storeDir, _ := os.Open(store.directory)
-	names, _ := storeDir.Readdirnames(n)
+	files, _ := storeDir.Readdir(-1)
+
+	// Sort the dentries.
+	sort.SliceStable(files, func(i, j int) bool {
```
Comment (harshavardhana) [inline pkg/event/target/queuestore.go]: No need to use Stable sorting @Praveenrajmani Stable sorts is a misnomer it is slower and not useful for us when our elements are not governing any pre-ordained order. 

https://github.com/minio/minio/pull/5401#discussion_r161648310

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: pkg/event/target/queuestore.go (hunk 1 above)
Comment (Praveenrajmani) [reply pkg/event/target/queuestore.go]: got it :+1:

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#2c6deb38-1409-47f2-a7ae-5ede84e6f506) |
<a name="2c6deb38-1409-47f2-a7ae-5ede84e6f506"></a>
### 7658-0ab4dc7/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:32399
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 39a3c5dadd25:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 11 minutes and 17 seconds
(4/14) Running aws-sdk-ruby tests ... done in 14 seconds
(5/14) Running awscli tests ... done in 1 minutes and 7 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 37 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 4 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 38 seconds
(10/14) Running minio-java tests ... FAILED in 7 minutes and 23 seconds
{
  "name": "minio-java",
  "function": "copyObject(String bucketName, String objectName, String destBucketName)",
  "duration": 300259,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code=XMinioReadQuorum, message=Multiple disk failures, unable to reconstruct data., bucketName=minio-java-test-54pdir, objectName=minio-java-test-1l9esa1, resource=/minio-java-test-54pdir/minio-java-test-1l9esa1, requestId=159F419D8BA65845, hostId=6fb05633-bc6f-421e-854b-aee73318566c)\nrequest={method=GET, url=http://72.28.97.52:32399/minio-java-test-54pdir/minio-java-test-1l9esa1, headers=Host: 72.28.97.52:32399\nUser-Agent: Minio (amd64; amd64) minio-java/dev\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20190516T194900Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20190516/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=503, headers=Accept-Ranges: bytes\nContent-Length: 1024\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"6c3f0d86af6d379b18f0c40adb69bffc-1\"\nLast-Modified: Thu, 16 May 2019 19:49:00 GMT\nRetry-After: 120\nServer: MinIO/DEVELOPMENT.2019-05-16T19-12-49Z\nVary: Origin\nX-Amz-Request-Id: 159F419D8BA65845\nX-Minio-Deployment-Id: 6fb05633-bc6f-421e-854b-aee73318566c\nX-Xss-Protection: 1; mode=block\nDate: Thu, 16 May 2019 19:49:01 GMT\n}\n >>> [io.minio.MinioClient.executeReq(MinioClient.java:1190), io.minio.MinioClient.execute(MinioClient.java:1056), io.minio.MinioClient.executeGet(MinioClient.java:1284), io.minio.MinioClient.getObject(MinioClient.java:1860), io.minio.MinioClient.getObject(MinioClient.java:1594), FunctionalTest.copyObject_test1(FunctionalTest.java:2017), FunctionalTest.runTests(FunctionalTest.java:3006), FunctionalTest.main(FunctionalTest.java:3121)]"
}

Executed 9 out of 14 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Review by nitisht — APPROVED** [review]
LGTM & Tested

**Review by nitisht — CHANGES_REQUESTED** [review]
Server starts when MQTT is not reachable and queueDir is not set, I think that is not expected

```
API: SYSTEM()
Time: 18:18:26 IST 05/21/2019
Error: Network Error : dial tcp 127.0.0.1:1883: connect: connection refused
       3: cmd/config-current.go:687:cmd.getNotificationTargets()
       2: cmd/notification.go:818:cmd.NewNotificationSys()
       1: cmd/server-main.go:362:cmd.serverMain()
Endpoint:  http://192.168.0.105:9000  http://172.17.0.1:9000  http://192.168.122.1:9000  http://192.168.42.1:9000  http://127.0.0.1:9000    
AccessKey: minio 
SecretKey: minio123 

Browser Access:
   http://192.168.0.105:9000  http://172.17.0.1:9000  http://192.168.122.1:9000  http://192.168.42.1:9000  http://127.0.0.1:9000    

Command-line Access: https://docs.min.io/docs/minio-client-quickstart-guide
   $ mc config host add myminio http://192.168.0.105:9000 minio minio123

Object API (Amazon S3 compatible):
   Go:         https://docs.min.io/docs/golang-client-quickstart-guide
   Java:       https://docs.min.io/docs/java-client-quickstart-guide
   Python:     https://docs.min.io/docs/python-client-quickstart-guide
   JavaScript: https://docs.min.io/docs/javascript-client-quickstart-guide
   .NET:       https://docs.min.io/docs/dotnet-client-quickstart-guide

```

**Comment by Praveenrajmani** [comment]
> Server starts when MQTT is not reachable and queueDir is not set, I think that is not expected
> 
> ```
> API: SYSTEM()
> Time: 18:18:26 IST 05/21/2019
> Error: Network Error : dial tcp 127.0.0.1:1883: connect: connection refused
>        3: cmd/config-current.go:687:cmd.getNotificationTargets()
>        2: cmd/notification.go:818:cmd.NewNotificationSys()
>        1: cmd/server-main.go:362:cmd.serverMain()
> Endpoint:  http://192.168.0.105:9000  http://172.17.0.1:9000  http://192.168.122.1:9000  http://192.168.42.1:9000  http://127.0.0.1:9000    
> AccessKey: minio 
> SecretKey: minio123 
> 
> Browser Access:
>    http://192.168.0.105:9000  http://172.17.0.1:9000  http://192.168.122.1:9000  http://192.168.42.1:9000  http://127.0.0.1:9000    
> 
> Command-line Access: https://docs.min.io/docs/minio-client-quickstart-guide
>    $ mc config host add myminio http://192.168.0.105:9000 minio minio123
> 
> Object API (Amazon S3 compatible):
>    Go:         https://docs.min.io/docs/golang-client-quickstart-guide
>    Java:       https://docs.min.io/docs/java-client-quickstart-guide
>    Python:     https://docs.min.io/docs/python-client-quickstart-guide
>    JavaScript: https://docs.min.io/docs/javascript-client-quickstart-guide
>    .NET:       https://docs.min.io/docs/dotnet-client-quickstart-guide
> ```
 
@nitisht The notification system itself works like that. This is a generic behavior across all the targets

- We cannot set an invalid config (`mc config set`) - when there is an inactive target
- But the server restarts/starts with an inactive target. Will print the error in the console though.

This was done in - #6126

**Comment by Praveenrajmani** [comment]
For reference - https://github.com/minio/minio/pull/6126

**Comment by nitisht** [comment]
> Note - The server will fail to start if the queueDir is not configured and the broker connection is inactive.

I was referring to this note in the PR description

**Comment by Praveenrajmani** [comment]
> > Note - The server will fail to start if the queueDir is not configured and the broker connection is inactive.
> 
> I was referring to this note in the PR description

Oh. Sorry.. it has to be `The server will fail to set the config if the queueDir is not configured and the broker connection is inactive.`.. will update the description

**Review by nitisht — APPROVED** [review]

**Review by sinhaashish — APPROVED** [review]
LGTM & Tested

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**